### ISSUE TITLE
Compiler segfault

### ISSUE THREAD

**Issue by MikeInnes** [ziglang/zig#21324]
### Zig Version

0.13.0

### Steps to Reproduce and Observed Behavior

The following code produces a segfault during compilation:

<details>

```zig
const std = @import("std");

fn down(comptime N: i64) i64 {
    // return if (N < 10) N + 1 else N;
    return N + 1;
}

pub fn VTable(comptime N: i64) type {
    return struct {
        open: *const fn (*anyopaque) Trie(down(N)),
        const Self = @This();
        pub fn init(_: type) Self {
            const impl = struct {
                pub fn open(_: *anyopaque) Trie(down(N)) {
                    unreachable;
                }
            };
            return Self{
                .open = impl.open,
            };
        }
    };
}

pub fn Trie(comptime N: i64) type {
    return struct {
        vtable: *const VTable(N),
    };
}

pub fn main() !void {
    _ = Trie(0){ .vtable = @ptrFromInt(8) };
}
```

</details>

```
➜  aot git:(aot) ✗ zig cc rt/bug.zig
[1]    63326 segmentation fault  zig cc rt/bug.zig
```

I'm using 0.13.0, but according to [godbolt](https://zig.godbolt.org/z/Tj93ax3zo) this seems to affect master.

### Expected Behavior

I imagine the code is invalid anyway, because Zig will try to instantiate the template for all `N`. If you enable the commented line to limit N, it compiles successfully. But in that case I'd expect a user-friendly error message explaining the problem.

**Comment by rohlem** [comment]
reproduced on zig version `0.14.0-dev.617+208baa37c`, reduced example:
```zig
pub fn Foo(comptime N: i64) type {
    return struct {
        foo: ?*Foo(N + 1) = null,
    };
}
test {
    _ = Foo(0){};
}
```

There just currently seems to be no limit on function instantiations via field types.
I think we should be able to just re-use `@setEvalBranchQuota` for this scenario, so code constructing too-far-nested types is required to raise the quota to do so.

**Comment by mlugg** [comment]
This is a surprisingly subtle issue which will require some (probably very boring) design work to give well-defined semantics for the eval branch quota here. I'll demonstrate this with two examples.

### We Can't Inherit The Quota For Type Resolution

Consider the following:
```zig
fn Foo() type {
    return struct {
        x: Bar(),
    };
}
const A = blk: {
    @setEvalBranchQuota(1000);
    break :blk Foo();
};
const B = blk: {
    @setEvalBranchQuota(10_000);
    break :blk Foo();
};
```
When `Foo` is initially called, no type resolution occurs. Eventually, some type resolution will be triggered which causes resolution of the field types of `Foo()` (i.e. evaluates the type expression `Bar()`). When this happens, what should the eval branch quota be? If it inherited 1000 or 10000 from `A`/`B`, we'd be making an arbitrary choice between the two, which we can't properly specify because analysis order is non-deterministic (we also might be resolving the type before some use sites are analyzed, so can't do the minimum or anything like that).

So then, the *only* possible choice here is to use a fresh quota of the default (1000). So, the expression `Bar()` is evaluated with a fresh eval branch quota of 1000.

### A Fresh Quota Gives Us Infinite Recursion

Okay, now consider rohlem's repro above:

```zig
pub fn Foo(comptime N: i64) type {
    return struct {
        foo: Foo(N + 1) = undefined,
    };
}
const A = Foo(0);
```

The eval branch quota when we call `Foo(0)` is 1000. It drops to 999 inside the call, but that's not relevant, because, by the above, we can't use this eval branch quota for anything else. We don't resolve the field types of `Foo(0)` immediately; this happens later on. When it does, we use a fresh eval branch quota, of the default of 1000. So, we then call `Foo(1)` with a quota of 1000, dropping to 999 within the call. By the same logic as above, later resolution will call `Foo(2)` with a quota of 1000, then `Foo(3)` with that same quota, etc, etc.

---

The crux of the problem here is that because container types are deduplicated (`Foo(0) == Foo(0)`), we don't have any "canonical" eval branch quota to inherit, so it is necessarily reset.

If I'm being honest, I have absolutely no idea how to solve this.

**Comment by rohlem** [comment]
@mlugg As long as we pick either 1_000 or 10_000, even if arbitrary (and inconsistent across runs), that still seems better than infinite recursion.
For usability it would be nice for the compile error to point to the quota it picked, so you know where in the code to raise it.

Alternatively, kind of an awkward workaround, but maybe this could be workable:
- introduce a new class of eval branch quota only for container field depth, and explicitly document that all but the first encountered value will be ignored.
  (Maybe that's enough reason that we give it a different name instead, like "field eval depth" or something?)
- allow `@setEvalBranchQuota`/`@setFieldEvalDepth` as a container-level declaration and add it as a `std.builtin.Type.Struct/Union` field for instantiated types.

**Comment by mlugg** [comment]
Non-determinism in whether a given piece of Zig code compiles is not acceptable.

Your alternative solution doesn't work for the same reason. There is no consistent "first encountered definition" by design.

**Comment by rohlem** [comment]
> Non-determinism in whether a given piece of Zig code compiles is not acceptable.

Then we could block the analysis, wait if other quotas show up through further analysis,
and only issue the error once we are certain that no satisfactory quota was supplied.
That way correct code will always compile.

**Comment by mlugg** [comment]
Here's a version of this that doesn't use type resolution at all. Instead, it uses the fact that declarations have independent eval quotas (again, a fundmental design choice which we won't be changing since it's necessary for concurrent analysis, incremental compilation, etc).
```zig
fn S(comptime x: u32) type {
    return struct {
        const val = S(x + 1).val;
    };
}

comptime {
    const T = S(0);
    _ = T.val;
}
```
I really don't like it, but I think a global quota might be the only solution to this, actually. It should be in *addition* to the eval branch quota, since the eval branch quota is useful for picking up common cases. The global quota could perhaps be set per-module and summed across all modules in the ZCU. cc @andrewrk and @SpexGuy to see if you can think of any cleaner solutions.

**Comment by SpexGuy** [comment]
This is what the instantiation quota is for. I'm having trouble finding any issues that reference it though, maybe that idea never made it to github?

The idea is that there's a separate limit to how many generic functions can be instantiated from any given context. Whenever a context triggers the instantiation of another context, e.g. in this case by referencing `T.val`, it adds one to the "stack" of instantiation. The instantiation quota is a depth limit for this stack. To make sure order-independence is preserved, every cached value needs to keep track of how deep its execution tree was, and uses of that value need to make sure they have enough quota to use the referenced value.

We'll probably want errors stemming from this to identify the cycle in the error message, but we don't need to keep that data during standard compilation. We can recompute it after the quota is violated.

Like the branch quota, there is a default value which is used for the root quota, and then code can modify it with a builtin.
